ui/src/app/views/workflowv3/graph/workflowv3-stages-graph.component.ts

```diff
@@ -0,0 +1,236 @@
+import {
+    AfterViewInit,
+    ChangeDetectionStrategy,
+    ChangeDetectorRef,
+    Component,
+    ComponentFactoryResolver,
+    ComponentRef,
+    EventEmitter,
+    HostListener,
+    Input,
+    OnDestroy,
+    Output,
+    ViewChild,
+    ViewContainerRef
+} from '@angular/core';
+import { AutoUnsubscribe } from 'app/shared/decorator/autoUnsubscribe';
+import { GraphNode, JobRun, WorkflowRunV3, WorkflowV3 } from '../workflowv3.model';
+import { WorkflowV3ForkJoinNodeComponent } from './workflowv3-fork-join-node.components';
+import { GraphDirection, WorkflowV3Graph } from './workflowv3-graph.lib';
+import { WorkflowV3JobNodeComponent } from './workflowv3-job-node.component';
+import { WorkflowV3JobsGraphComponent } from './workflowv3-jobs-graph.component';
+
+export type WorkflowV3JobsGraphOrNodeComponent = WorkflowV3JobsGraphComponent |
+    WorkflowV3ForkJoinNodeComponent | WorkflowV3JobNodeComponent;
+
+@Component({
+    selector: 'app-workflowv3-stages-graph',
+    templateUrl: './workflowv3-stages-graph.html',
+    styleUrls: ['./workflowv3-stages-graph.scss'],
+    changeDetection: ChangeDetectionStrategy.OnPush
+})
+@AutoUnsubscribe()
+export class WorkflowV3StagesGraphComponent implements AfterViewInit, OnDestroy {
+    static maxScale = 15;
+    static minScale = 1 / 5;
+
+    nodes: Array<GraphNode> = [];
+    @Input() set workflow(data: WorkflowV3) {
+        this.nodes = [];
+        this.hasStages = !!data && !!data.stages;
+        if (data && data.stages) {
+            this.nodes.push(...Object.keys(data.stages)
+                .map(k => <GraphNode>{ name: k, depends_on: data.stages[k].depends_on, sub_graph: [] }));
+        }
+        if (data && data.jobs) {
+            Object.keys(data.jobs).forEach(k => {
+                let j = data.jobs[k];
+                let node = <GraphNode>{ name: k, depends_on: j.depends_on };
+                if (this.jobRuns[k]) {
+                    node.run = this.jobRuns[k][0];
+                }
+                if (j.stage) {
+                    for (let i = 0; i < this.nodes.length; i++) {
+                        if (this.nodes[i].name === j.stage) {
+                            this.nodes[i].sub_graph.push(node);
+                            break;
+                        }
+                    }
+                } else {
+                    this.nodes.push(node);
+                }
+            });
+        }
+        this.changeDisplay();
+    }
+
+    jobRuns: { [name: string]: Array<JobRun> } = {};
+    @Input() set workflowRun(data: WorkflowRunV3) {
+        if (!data) {
+            return;
```

Comment:
delete this.jobRuns

Reply:
To improve when adding navigation between runs.